### PR TITLE
MAINT: Clarify the error message for resize failure

### PR DESCRIPTION
**Pull request by kngwyu**
Curerntly 
```python
>>> arr = np.array([1, 2, 3])
>>> arr_ref = arr
>>> arr.resize(100)
```
causes this error.
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: cannot resize an array that references or is referenced
by another array in this way.  Use the resize function
```
But, I feel it odd to tell users to use `resize`, when `resize` fails.

### CODE REVIEW

**Review by eric-wieser — CHANGES_REQUESTED** [review]
The wording is "use the resize _function_", not "use the resize _method_". Perhaps it would be clearer to say `use the np.resize function` 

`reshape` is not a substitute for `resize`.

**Comment by kngwyu** [comment]
Thanks, I changed to just propose `np.resize` or `refcheck=False`.

File: numpy/core/src/multiarray/shape.c [hunk 1]
```diff
@@ -108,7 +108,7 @@ PyArray_Resize(PyArrayObject *self, PyArray_Dims *newshape, int refcheck,
             PyErr_SetString(PyExc_ValueError,
                     "cannot resize an array that "
                     "references or is referenced\n"
-                    "by another array in this way.  Use the resize function");
+                    "by another array in this way.  Use the np.resize function or refcheck=False");
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/shape.c]: This is not correct - passing `refcheck = False` doesn't help if `PyArray_BASE(arr)` is set.

Either:
* remove the remark about refcheck
* Split this `if` into two ifs, and only mention refcheck in the one where `refcnt > 2` is true.

**Comment by kngwyu** [comment]
Thanks, fixed by the later way.

File: numpy/core/src/multiarray/shape.c [hunk 2]
```diff
@@ -102,13 +102,20 @@ PyArray_Resize(PyArrayObject *self, PyArray_Dims *newshape, int refcheck,
...
+            return NULL;
+        }
+        if (PyArray_BASE(self) != NULL
+              || (((PyArrayObject_fields *)self)->weakreflist != NULL)) {
+            PyErr_SetString(PyExc_ValueError,
+                    "cannot resize an array that "
+                    "references or is referenced\n"
+                    "by another array in this way. Use the np.resize function.");
             return NULL;
         }
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/shape.c]: As a follow up, you could move this check to be done before the other two. As you have it, there are cases where the user sees "A is not allowed, try B or C", then tries B and gets "B is not allowed, try C". If you move this check first, then the user will see "A is not allowed, try C", and fix it in one step.

**Review by eric-wieser — APPROVED** [review]

**Comment by eric-wieser** [comment]
Would be great to add a test for that code path that codecov complains about. Creating a `weakref` to the array then trying to resize it should do the trick.

**Comment by kngwyu** [comment]
Thanks. Please let me know if test location is wrong.

**Review by eric-wieser — APPROVED** [review]
Looks good assuming tests pass

**Comment by mattip** [comment]
Thanks @kngwyu

**Comment by charris** [comment]
Fails for wheel tests on Python 3.7. See [here.](https://travis-ci.org/numpy/numpy/jobs/612498344?utm_medium=notification&utm_source=github_status)

**Pull request merged**